### ISSUE TITLE
Importer always runs out of disk space

### ISSUE THREAD

**Issue by timstoop** [kubevirt/containerized-data-importer#1078]
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:
When downloading a raw image from S3, the importer always seems to run out of disk space after the download completed.

**What you expected to happen**:
I expected a disk more or less the same size as the original disk (from which the raw image was created) to be large enough to transfer the data to.

**How to reproduce it (as minimally and precisely as possible)**:
Create a raw image from a disk with a known size. Mine looked like this: 
```
-rw-r--r-- 1 tim tim 26843545600 Jan  6 13:21 app1.raw
```
Create a DV a little larger than this:
```
apiVersion: cdi.kubevirt.io/v1alpha1
kind: DataVolume
metadata:
  name: "app1"
spec:
  source:
      http:
         url: "https://s3-eu-west-1.amazonaws.com/very-secret-bucket/app1.raw"
  pvc:
    accessModes:
      - ReadWriteOnce
    resources:
      requests:
        storage: "30Gi"
    storageClassName: ssd
```
The importer starts the download and ends with an error:
```
I0111 12:12:59.597799       1 importer.go:51] Starting importer
I0111 12:12:59.598514       1 importer.go:107] begin import process
I0111 12:12:59.765095       1 data-processor.go:253] Calculating available size
I0111 12:12:59.768593       1 data-processor.go:261] Checking out file system volume size.
I0111 12:12:59.768623       1 data-processor.go:265] Request image size not empty.
I0111 12:12:59.768637       1 data-processor.go:270] Target size 31509590016.
I0111 12:12:59.768734       1 util.go:37] deleting file: /data/lost+found
I0111 12:12:59.773517       1 data-processor.go:183] New phase: Convert
I0111 12:12:59.773541       1 data-processor.go:189] Validating image
I0111 12:13:00.020635       1 qemu.go:212] 0.00
I0111 12:13:02.246148       1 qemu.go:212] 1.01
I0111 12:13:04.415218       1 qemu.go:212] 2.02
.
.
.
I0111 12:16:22.752447       1 qemu.go:212] 97.76
I0111 12:16:24.671371       1 qemu.go:212] 98.77
I0111 12:16:26.933390       1 qemu.go:212] 99.77
I0111 12:16:27.529702       1 data-processor.go:183] New phase: Resize
W0111 12:16:27.582637       1 data-processor.go:240] Available space less than requested size, resizing image to available space 31509590016.
I0111 12:16:27.582936       1 data-processor.go:246] Expanding image size to: 31509590016
E0111 12:16:27.621228       1 prlimit.go:164] qemu-img failed output is:
E0111 12:16:27.621509       1 prlimit.go:165] qemu-img: Could not preallocate new data: No space left on device

E0111 12:16:27.637249       1 prlimit.go:164] qemu-img failed output is:
E0111 12:16:27.637272       1 prlimit.go:165] qemu-img: Could not preallocate new data: No space left on device

E0111 12:16:27.651494       1 prlimit.go:164] qemu-img failed output is:
E0111 12:16:27.651514       1 prlimit.go:165] qemu-img: Could not preallocate new data: No space left on device

I0111 12:16:27.819595       1 data-processor.go:183] New phase: Complete
I0111 12:16:27.842616       1 importer.go:160] Import complete
```
**Anything else we need to know?**:
I tried the same on a 50Gi image, same result.

**Environment**:
- CDI version (use `kubectl get deployments cdi-deployment -o yaml`): v1.11.0
- Kubernetes version (use `kubectl version`): 1.15.7
- Cloud provider or hardware configuration: GCE
- Install tools: ?
- Others: ?

**Comment by mhenriks** [comment]
Hi @timstoop it looks like the file was successfully downloaded but the virtual size (31509590016) is larger than the physical size so qemu-img attempts to preallocate that additional space on the filesystem.  That operation is failing because there is not enough free space on the filesystem.  I assume because of filesystem overhead.  You can confirm by making a 30Gi PVC on the same storageclass, and execing into dummy pod that has the pvc mounted to check free space.

So, I'd try making the PVC bigger.

**Comment by timstoop** [comment]
I tried the same with a 50Gi PVC, same result.

**Comment by timstoop** [comment]
Just to show you, we're still talking about the same 25Gi RAW image:

```
I0112 14:16:12.853925       1 qemu.go:212] 96.75
I0112 14:16:14.846172       1 qemu.go:212] 97.76
I0112 14:16:16.920670       1 qemu.go:212] 98.77
I0112 14:16:18.844510       1 qemu.go:212] 99.77
I0112 14:16:19.386160       1 data-processor.go:183] New phase: Resize
W0112 14:16:19.800593       1 data-processor.go:240] Available space less than requested size, resizing image to available space 52504764416.
I0112 14:16:19.801253       1 data-processor.go:246] Expanding image size to: 52504764416
E0112 14:16:19.963950       1 prlimit.go:164] qemu-img failed output is:
E0112 14:16:19.964236       1 prlimit.go:165] qemu-img: Could not preallocate new data: No space left on device

E0112 14:16:19.994516       1 prlimit.go:164] qemu-img failed output is:
E0112 14:16:19.994552       1 prlimit.go:165] qemu-img: Could not preallocate new data: No space left on device

E0112 14:16:20.025315       1 prlimit.go:164] qemu-img failed output is:
E0112 14:16:20.025337       1 prlimit.go:165] qemu-img: Could not preallocate new data: No space left on device

I0112 14:16:20.198182       1 data-processor.go:183] New phase: Complete
I0112 14:16:20.198361       1 importer.go:160] Import complete
```

**Comment by timstoop** [comment]
I'm kind of wondering, what is it trying to expand? Because the raw image is a dd of a disk, which means it starts with a GPT and then continues with partitions which contain an LVM PV, on top of which there are VGs and LVs. What is it trying to expand?

**Comment by timstoop** [comment]
Latest results are even weirder:

```
I0112 20:59:50.344502       1 qemu.go:212] 98.38                                          
I0112 20:59:54.495058       1 qemu.go:212] 99.39                            
I0112 20:59:58.948830       1 data-processor.go:183] New phase: Resize                    
W0112 20:59:59.007585       1 data-processor.go:240] Available space less than requested size, resizing image to available space 68089708544.
I0112 20:59:59.007906       1 data-processor.go:246] Expanding image size to: 68089708544 
E0112 20:59:59.103376       1 prlimit.go:164] qemu-img failed output is:
E0112 20:59:59.103680       1 prlimit.go:165] qemu-img: Could not preallocate new data: No space left on device
                                                                             
E0112 20:59:59.126814       1 prlimit.go:164] qemu-img failed output is:   
E0112 20:59:59.126838       1 prlimit.go:165] qemu-img: Could not preallocate new data: No space left on device
                                                                                
E0112 20:59:59.149021       1 prlimit.go:164] qemu-img failed output is:    
E0112 20:59:59.149049       1 prlimit.go:165] qemu-img: Could not preallocate new data: No space left on device
                                                                                          
E0112 20:59:59.170847       1 prlimit.go:164] qemu-img failed output is:    
E0112 20:59:59.170868       1 prlimit.go:165] qemu-img: Could not preallocate new data: No space left on device
                                                                                 
E0112 20:59:59.193058       1 prlimit.go:164] qemu-img failed output is:                  
E0112 20:59:59.193079       1 prlimit.go:165] qemu-img: Could not preallocate new data: No space left on device
                                                                                          
E0112 20:59:59.215095       1 prlimit.go:164] qemu-img failed output is:    
E0112 20:59:59.215118       1 prlimit.go:165] qemu-img: Could not preallocate new data: No space left on device
                                                                                 
E0112 20:59:59.237895       1 prlimit.go:164] qemu-img failed output is:                  
E0112 20:59:59.237920       1 prlimit.go:165] qemu-img: Could not preallocate new data: No space left on device
                                                                                        
E0112 20:59:59.261722       1 prlimit.go:164] qemu-img failed output is:                  
E0112 20:59:59.261744       1 prlimit.go:165] qemu-img: Could not preallocate new data: No space left on device
                                                                                          
E0112 20:59:59.284747       1 prlimit.go:164] qemu-img failed output is:        
E0112 20:59:59.284771       1 prlimit.go:165] qemu-img: Could not preallocate new data: No space left on device
                                                                            
E0112 20:59:59.308688       1 prlimit.go:164] qemu-img failed output is:                  
E0112 20:59:59.308720       1 prlimit.go:165] qemu-img: Could not preallocate new data: No space left on device
                                                                                          
E0112 20:59:59.332524       1 prlimit.go:164] qemu-img failed output is:
E0112 20:59:59.332548       1 prlimit.go:165] qemu-img: Could not preallocate new data: No space left on device
                                                                                          
E0112 20:59:59.332577       1 data-processor.go:180] exit status 1         
qemu-img execution failed                                                               
kubevirt.io/containerized-data-importer/pkg/system.ExecWithLimits               
        /go/src/kubevirt.io/containerized-data-importer/pkg/system/prlimit.go:166
kubevirt.io/containerized-data-importer/pkg/image.(*qemuOperations).Resize       
        /go/src/kubevirt.io/containerized-data-importer/pkg/image/qemu.go:139             
kubevirt.io/containerized-data-importer/pkg/image.QEMUOperations.Resize-fm
        /go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:247
kubevirt.io/containerized-data-importer/pkg/util.RetryBackoffSize
        /go/src/kubevirt.io/containerized-data-importer/pkg/util/util.go:142     
kubevirt.io/containerized-data-importer/pkg/importer.ResizeImage          
        /go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:247
kubevirt.io/containerized-data-importer/pkg/importer.(*DataProcessor).resize
        /go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:217
kubevirt.io/containerized-data-importer/pkg/importer.(*DataProcessor).ProcessData
        /go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:172
main.main
        /go/src/kubevirt.io/containerized-data-importer/cmd/cdi-importer/importer.go:142
runtime.main
        /gimme/.gimme/versions/go1.11.5.linux.amd64/src/runtime/proc.go:201
runtime.goexit
        /gimme/.gimme/versions/go1.11.5.linux.amd64/src/runtime/asm_amd64.s:1333
Error resizing image /data/disk.img
kubevirt.io/containerized-data-importer/pkg/image.(*qemuOperations).Resize
        /go/src/kubevirt.io/containerized-data-importer/pkg/image/qemu.go:141
kubevirt.io/containerized-data-importer/pkg/image.QEMUOperations.Resize-fm
        /go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:247
kubevirt.io/containerized-data-importer/pkg/util.RetryBackoffSize
        /go/src/kubevirt.io/containerized-data-importer/pkg/util/util.go:142
kubevirt.io/containerized-data-importer/pkg/importer.ResizeImage
        /go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:247
kubevirt.io/containerized-data-importer/pkg/importer.(*DataProcessor).resize
        /go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:217
kubevirt.io/containerized-data-importer/pkg/importer.(*DataProcessor).ProcessData
        /go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:172
main.main
        /go/src/kubevirt.io/containerized-data-importer/cmd/cdi-importer/importer.go:142
runtime.main
        /gimme/.gimme/versions/go1.11.5.linux.amd64/src/runtime/proc.go:201
runtime.goexit
        /gimme/.gimme/versions/go1.11.5.linux.amd64/src/runtime/asm_amd64.s:1333
Resize of image failed
kubevirt.io/containerized-data-importer/pkg/importer.(*DataProcessor).resize
        /go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:219
kubevirt.io/containerized-data-importer/pkg/importer.(*DataProcessor).ProcessData
        /go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:172
main.main
        /go/src/kubevirt.io/containerized-data-importer/cmd/cdi-importer/importer.go:142
runtime.main
        /gimme/.gimme/versions/go1.11.5.linux.amd64/src/runtime/proc.go:201
runtime.goexit
        /gimme/.gimme/versions/go1.11.5.linux.amd64/src/runtime/asm_amd64.s:1333
Unable to resize disk image to requested size
kubevirt.io/containerized-data-importer/pkg/importer.(*DataProcessor).ProcessData
        /go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:174
main.main
        /go/src/kubevirt.io/containerized-data-importer/cmd/cdi-importer/importer.go:142
runtime.main
        /gimme/.gimme/versions/go1.11.5.linux.amd64/src/runtime/proc.go:201
runtime.goexit
        /gimme/.gimme/versions/go1.11.5.linux.amd64/src/runtime/asm_amd64.s:1333
E0112 20:59:59.332707       1 importer.go:144] exit status 1
qemu-img execution failed
kubevirt.io/containerized-data-importer/pkg/system.ExecWithLimits
        /go/src/kubevirt.io/containerized-data-importer/pkg/system/prlimit.go:166
kubevirt.io/containerized-data-importer/pkg/image.(*qemuOperations).Resize
        /go/src/kubevirt.io/containerized-data-importer/pkg/image/qemu.go:139
kubevirt.io/containerized-data-importer/pkg/image.QEMUOperations.Resize-fm
        /go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:247
kubevirt.io/containerized-data-importer/pkg/util.RetryBackoffSize
        /go/src/kubevirt.io/containerized-data-importer/pkg/util/util.go:142
kubevirt.io/containerized-data-importer/pkg/importer.ResizeImage
        /go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:247
kubevirt.io/containerized-data-importer/pkg/importer.(*DataProcessor).resize
        /go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:217
kubevirt.io/containerized-data-importer/pkg/importer.(*DataProcessor).ProcessData
        /go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:172
main.main
        /go/src/kubevirt.io/containerized-data-importer/cmd/cdi-importer/importer.go:142
runtime.main
        /gimme/.gimme/versions/go1.11.5.linux.amd64/src/runtime/proc.go:201
runtime.goexit
        /gimme/.gimme/versions/go1.11.5.linux.amd64/src/runtime/asm_amd64.s:1333
Error resizing image /data/disk.img
kubevirt.io/containerized-data-importer/pkg/image.(*qemuOperations).Resize
        /go/src/kubevirt.io/containerized-data-importer/pkg/image/qemu.go:141
kubevirt.io/containerized-data-importer/pkg/image.QEMUOperations.Resize-fm
        /go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:247
kubevirt.io/containerized-data-importer/pkg/util.RetryBackoffSize
        /go/src/kubevirt.io/containerized-data-importer/pkg/util/util.go:142
kubevirt.io/containerized-data-importer/pkg/importer.ResizeImage
        /go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:247
kubevirt.io/containerized-data-importer/pkg/importer.(*DataProcessor).resize
        /go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:217
kubevirt.io/containerized-data-importer/pkg/importer.(*DataProcessor).ProcessData
        /go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:172
main.main
        /go/src/kubevirt.io/containerized-data-importer/cmd/cdi-importer/importer.go:142
runtime.main
        /gimme/.gimme/versions/go1.11.5.linux.amd64/src/runtime/proc.go:201
runtime.goexit
        /gimme/.gimme/versions/go1.11.5.linux.amd64/src/runtime/asm_amd64.s:1333
Resize of image failed
kubevirt.io/containerized-data-importer/pkg/importer.(*DataProcessor).resize
        /go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:219
kubevirt.io/containerized-data-importer/pkg/importer.(*DataProcessor).ProcessData
        /go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:172
main.main
        /go/src/kubevirt.io/containerized-data-importer/cmd/cdi-importer/importer.go:142
runtime.main
        /gimme/.gimme/versions/go1.11.5.linux.amd64/src/runtime/proc.go:201
runtime.goexit
        /gimme/.gimme/versions/go1.11.5.linux.amd64/src/runtime/asm_amd64.s:1333
Unable to resize disk image to requested size
kubevirt.io/containerized-data-importer/pkg/importer.(*DataProcessor).ProcessData
        /go/src/kubevirt.io/containerized-data-importer/pkg/importer/data-processor.go:174
main.main
        /go/src/kubevirt.io/containerized-data-importer/cmd/cdi-importer/importer.go:142
runtime.main
        /gimme/.gimme/versions/go1.11.5.linux.amd64/src/runtime/proc.go:201
runtime.goexit
        /gimme/.gimme/versions/go1.11.5.linux.amd64/src/runtime/asm_amd64.s:1333
```

I'm not sure what changed, I did not upgrade anything.

**Comment by mhenriks** [comment]
attn @awels 

@timstoop I think I better understand what is going on here.  CDI is attempting to expand the image so you have less unused space on your PVC.

What provisioner/filesystem are you using?

**Comment by awels** [comment]
@mhenriks So the logs are showing the logic that is trying to expand the disk image to fill the PVC size to its maximum capacity. Unfortunately different file systems/provisioners have different limits, so we resize to progressively smaller max sizes until it either succeeds or we hit the retry limit. To give an example. Lets say you create a 50G PVC, because your provisioner creates a block device by default, it will put a file system on there. The file system reports It has 495653454096 bytes of available space. So we attempt to resize the disk image to exactly that size. This fails because of variable overhead by whatever file system is on the block device (xfs for instance exhibits this behaviour). We see this failure and print: qemu-img: Could not preallocate new data: No space left on device. Now we retry but with slightly smaller size (25 block of 4k smaller to be exact). Until we either succeed or hit the retry limit (10 retries).

Now in the original report, it looks like the import actually succeeded, it just needed a few retries to find a size that worked. So basically it worked, our messages are unclear obviously.

So in a nutshell what should exist after it is done resizing. Lets say you started with a 25G image that has a 25G partition (and file system) on it. And you created a 30G PVC, after it finishes resizing you will have a 'close to 30G' disk image, with a 25G partition (and file system) on it. So you will have 5G of available raw disk you can create another partition on, or resize your existing partition to fill up the space. 

We are not currently doing that last step for various reasons, but you could for instance on first boot run something to resize the partition to fill the space.

**Comment by mhenriks** [comment]
This should be fixed as CDI no longer attempts to expand images
